src/main/scala/util/Property.scala

```diff
@@ -0,0 +1,45 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util.property
+
+import Chisel._
+import chisel3.internal.sourceinfo.{SourceInfo, SourceLine}
+import freechips.rocketchip.config.{Field, Parameters}
+
+case object PropertyLibrary extends Field[BasePropertyLibrary[BasePropertyParameters]]
+
+
+sealed abstract class PropertyType(name: String) {
+  override def toString: String = name
+}
+
+object PropertyType {
+  object Assert extends PropertyType("Assert")
+  object Assume extends PropertyType("Assume")
+  object Cover extends PropertyType("Cover")
+}
+
+class BasePropertyParameters(val p_type: PropertyType, val cond: Bool, val label: String, val message: String) {}
+
+class BasePropertyLibrary[PropParam <: BasePropertyParameters] {
+  def generateCover(prop_param: PropParam)(implicit sourceInfo: SourceInfo): Unit = {
+    // Currently empty for user to extend their own cover methods
+    Unit
+  }
+}
+
+object cover {
+  def apply(cond: Bool)(implicit sourceInfo: SourceInfo, p: Parameters): Unit = {
+    val prop_param = new BasePropertyParameters(PropertyType.Cover, cond, "", "")
+    p(PropertyLibrary).generateCover(prop_param)
+  }
+  def apply(cond: Bool, label: String)(implicit sourceInfo: SourceInfo, p: Parameters): Unit = {
+    val prop_param = new BasePropertyParameters(PropertyType.Cover, cond, label, "")
+    p(PropertyLibrary).generateCover(prop_param)
+  }
+  def apply(cond: Bool, label: String, message: String)(implicit sourceInfo: SourceInfo, p: Parameters): Unit = {
+    val prop_param = new BasePropertyParameters(PropertyType.Cover, cond, label, message)
+    p(PropertyLibrary).generateCover(prop_param)
+  }
+}
```

Comment:
If all `PropertyLibrary` child classes can agree that `generateCover` takes a `CoverPropertyParamers` then you can simplify this to the following
:
```
object cover {
  def apply(cond: Bool)(implicit sourceInfo: SourceInfo, p: Parameters) {
    p(PropertyLibrary).generateCover(CoverPropertyParamers(cond))
  }
  def apply(cond: Bool, label: String)(implicit sourceInfo: SourceInfo, p: Parameters) {
    p(PropertyLibrary).generateCover(CoverPropertyParamers(cond, label))
  }
  def apply(cond: Bool, label: String, message: String)(implicit sourceInfo: SourceInfo, p: Parameters) {
    p(PropertyLibrary).generateCover(CoverPropertyParamers(cond, label, message))
  }
}
```